ci/test/build.py

```diff
@@ -16,7 +16,7 @@
 
 
 def main() -> None:
-    repo = mzbuild.Repository(Path("."))
+    repo = mzbuild.Repository(Path("."), coverage=True)
```

Comment:
@benesch would be the authoritative source on how to structure this, but I think `coverage` should default to `False` across the board so that we do not end up publishing coverage-enabled containers as the official release on the cloud.

I think it should work that way:
1. Keep the existing build and CI steps as they are
2. Create a separate Buildkite pipeline. We can trigger that pipeline manually for the time being, later it can be set to trigger in the same circumstances as the main Tests pipeline.
3. In the Buildkite UI, that pipeline should perform a single step:
```
steps:
  - command: "buildkite-agent pipeline upload ci/coverage/pipeline.yml"
    priority: 2
    agents:
      queue: "*"
```

Then, `ci/coverage/pipeline.yml` can contain the following steps:

```
env:
  CI_BUILDER_SCCACHE: 1
  CI_COVERAGE_ENABLED: 1

steps:
  - id: build-x86_64-coverage
    label: Build x86_64 with coverage 
    command: bin/ci-builder run stable bin/pyactivate -m ci.test.build x86_64 --coverage
    inputs:
      - "*"
    timeout_in_minutes: 60
    priority: 1
    agents:
      queue: builder-linux-x86_64
 - id: mkpipeline
     bin/ci-builder run stable bin/pyactivate -m ci.test.mkpipeline

  - id: coverage-pr-analyze
    label: Analyze code coverage for PR
    command: bin/ci-builder run stable ci/test/coverage_report.sh
    inputs: ["*"]
    priority: 1
    agents:
      queue: linux-x86_64
```
(build.py should take a `--coverage` argument. I am not even sure where the `x86_64` argument is parsed and processed)
(mkpipeline.py should just do the same thing that it does in the default pipeline -- basically put the steps from `ci/test/pipeline.template.yml` into the coverage pipeline via `buildkite-agent pipeline upload`. The only trick here is that the `coverage-pr-analyze` step needs to survive the `pipeline upload` call (that is, be present in the replacement pipeline as well). )

4. mzcompose should look for the presence of the `CI_COVERAGE_ENABLED` env variable and use coverage-enabled images accordingly. Mzbuild.py already includes a line `self_hash.update(f"coverage={self.image.rd.coverage}".encode())` , so the only thing that is needed is to make sure self.image.rd.coverage = True when the image is being built and when the image is being used by mzcompose.

5.  All coverage-related actions in the Buildkite plugins should also only happen if `CI_COVERAGE_ENABLED` is enabled.
I hope this helps. Let me know if a Zoom call would be helpful.

Reply:
@philip-stoev's proposal sounds about right to me, but I only had time to skim quickly. Definitely we don't want to be uploading coverage-enabled builds by default. 

Have we run the math on how much this will cost? Running CI twice in order to get coverage information sounds like it could be quite expensive.

Reply:
How about only running coverage CI on PRs, and only non-coverage on main branch? I also thought that a full additional coverage run for each PR would be too expensive. If we trigger it manually for just a few changes, it would probably be ok, but of more limited value. Could be fine to start with.

Reply:
The problem with running with coverage by default in PR is that any timing-related issues that are triggered that way will not be easily reproducible on a developer or scratch machine, unless the developer also requests mzcompose to use coverage binaries locally. We already have issues with the reproducibility of some issues that are only seen in the CI so we should not do anything that exacerbates the problem.

How about this as a cost-cutting measure:
1. Have the coverage as a separate pipeline still.
2. By default, it only runs testdrive , sqllogictest and even possibly uses a glob to run any other CI jobs whose test files have changed (that is, if a new python test is added by the PR, it will be run as it is assumed that it contributes to the coverage of the feature at hand)
3. If that produces a report with obvious coverage holes, a member of the QA team will review it. They may then trigger a coverage build over the entire set of available tests to allow those tests to close any coverage gaps.
(as this will be a cost optimization, it can be tackled as a separate task)